File: Csharp/MPP/protobuff/ProtoBaskWorker.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Net.Sockets;
using System.Text;
using System.Threading;
using System.Threading.Tasks;
using Google.Protobuf;
using Proto;
using model2;
using services2;

namespace protobuf
{
    public class ProtoBaskWorker : IBasketObserver
    {
        private IBasketService server;
        private TcpClient connection;

        private NetworkStream stream;
        private volatile bool connected;

        public ProtoBaskWorker(IBasketService server, TcpClient connection)
        {
            this.server = server;
            this.connection = connection;
            try
            {
                stream = connection.GetStream();
				Console.WriteLine("ASta ii streamu");
				Console.WriteLine(stream);
                connected = true;
            }
            catch (Exception e)
            {
                Console.WriteLine(e.StackTrace);
            }
        }

		public virtual void run()
		{
			while (connected)
			{
				try
				{
					Console.WriteLine("Getting request");
					Console.WriteLine(stream.ToString());
					BasketRequest request = BasketRequest.Parser.ParseDelimitedFrom(stream);
					Console.WriteLine("Not passing request");
					BasketResponse response = handleRequest(request);
					if (response != null)
					{
						sendResponse(response);
					}
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}

				try
				{
					Thread.Sleep(1000);
				}
				catch (Exception e)
				{
					Console.WriteLine(e.StackTrace);
				}
			}
			try
			{
				stream.Close();
				connection.Close();
			}
			catch (Exception e)
			{
				Console.WriteLine("Error " + e);
			}
		}

        private BasketResponse handleRequest(BasketRequest request)
        {
			BasketResponse response = null;
			BasketRequest.Types.Type reqType = request.Type;
			switch (reqType)
			{
				case BasketRequest.Types.Type.Login:
					{
						Console.WriteLine("Login request ...");
						model2.Organiser user = ProtoUtils.getUser(request);
						try
						{
							lock (server)
							{
								server.login(user, this);
							}
							return ProtoUtils.createOkResponse();
						}
						catch (BasketException e)
						{
							connected = false;
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
				case BasketRequest.Types.Type.Logout:
					{
						Console.WriteLine("Logout request");
						model2.Organiser user = ProtoUtils.getUser(request);
						try
						{
							lock (server)
							{

								server.logout(user, this);
							}
							connected = false;
							return ProtoUtils.createOkResponse();

						}
						catch (BasketException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
				case BasketRequest.Types.Type.GetMatches:
					{
						Console.WriteLine("Get matches request ...");
						try
						{
							model2.Match[] matches;
							lock (server)
							{
								matches = server.getMatches();
							}
							return ProtoUtils.createGetMatchesResponse(matches);
						}
						catch (BasketException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}

				case BasketRequest.Types.Type.UpdateMatches:
					{
						Console.WriteLine("Updating matches request ...");
						model2.Ticket ticket = ProtoUtils.getTicket(request);  //DTOUtils.getFromDTO(udto);
						try
						{
							model2.Match[] matches;
							lock (server)
							{

								server.sendUpdate(ticket);
							}
							return ProtoUtils.createOkResponse();
						}
						catch (BasketException e)
						{
							return ProtoUtils.createErrorResponse(e.Message);
						}
					}
			}
			return response;
		}

        private void sendResponse(BasketResponse response)
        {
			Console.WriteLine("sending response " + response);
			lock (stream)
			{
				response.WriteDelimitedTo(stream);
				stream.Flush();
			}
		}

        public void updateReceived(model2.Match[] matches)
        {
			Console.WriteLine("Match list updated");
			try
			{
				sendResponse(ProtoUtils.createUpdateMatchesResponse(matches));
			}
			catch (Exception e)
			{
				Console.WriteLine(e.StackTrace);
			}
		}
    }
}

File: JavaSockets/BaskNetworking/src/main/java/bask/network/protobuffprotocol/ProtoBaskWorker.java
package bask.network.protobuffprotocol;

import bask.model.Match;
import bask.model.Organiser;
import bask.model.Ticket;
import bask.services.BasketException;
import bask.services.IObserver;
import bask.services.IServices;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.net.Socket;
import java.util.List;

import static bask.network.protobuffprotocol.BasketProtobufs.BasketResponse.Type.NewMatchList;

public class ProtoBaskWorker  implements Runnable, IObserver {
    private IServices server;
    private Socket connection;

    private InputStream input;
    private OutputStream output;
    private volatile boolean connected;

    public ProtoBaskWorker(IServices server, Socket connection) {
        this.server = server;
        this.connection = connection;
        try{
            output=connection.getOutputStream() ;//new ObjectOutputStream(connection.getOutputStream());
            input=connection.getInputStream(); //new ObjectInputStream(connection.getInputStream());
            connected=true;
        } catch (IOException e) {
            e.printStackTrace();
        }
    }


    @Override
    public void listUpdated(List<Match> matches) throws BasketException {
        System.out.println("Match list updated");
        Match[] matchesArr = new Match[matches.size()];
        matches.toArray(matchesArr);
        try {
            sendResponse(ProtoUtils.createUpdateMatchesResponse(matchesArr));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void organiserLoggedIn(Organiser organiser) throws BasketException {
        System.out.println("Organiser logged in "+organiser);
        try {
            sendResponse(ProtoUtils.createLoggedInResponse(organiser));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void organiserLoggedOut(Organiser organiser) throws BasketException {
        System.out.println("Organiser logged out "+organiser);
        try {
            sendResponse(ProtoUtils.createLoggedOutResponse(organiser));
        } catch (IOException e) {
            e.printStackTrace();
        }
    }

    @Override
    public void run() {
        while(connected){
            try {
                // Object request=input.readObject();
                System.out.println("Waiting requests ...");
                BasketProtobufs.BasketRequest request=BasketProtobufs.BasketRequest.parseDelimitedFrom(input);
                System.out.println("Request received: "+request);
                BasketProtobufs.BasketResponse response=handleRequest(request);
                if (response!=null){
                    sendResponse(response);
                }
            } catch (IOException e) {
                e.printStackTrace();
            }
            try {
                Thread.sleep(1000);
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        try {
            input.close();
            output.close();
            connection.close();
        } catch (IOException e) {
            System.out.println("Error "+e);
        }
    }

    private BasketProtobufs.BasketResponse handleRequest(BasketProtobufs.BasketRequest request) {
        BasketProtobufs.BasketResponse response=null;
        switch (request.getType()){
            case Login:{
                System.out.println("Login request ...");
                Organiser user=ProtoUtils.getUser(request);
                try {
                    server.login(user, this);
                    return ProtoUtils.createOkResponse();
                } catch (BasketException e) {
                    connected=false;
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case Logout:{
                System.out.println("Logout request");
                Organiser user=ProtoUtils.getUser(request);
                try {
                    server.logout(user, this);
                    connected=false;
                    return ProtoUtils.createOkResponse();

                } catch (BasketException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case GetMatches:{
                System.out.println("Get matches Request ...");
                try {
                    List<Match> matches = server.getMatchesList();
                    Match[] matchesArr = new Match[matches.size()];
                    matches.toArray(matchesArr);
                    return ProtoUtils.createGetMatchesResponse(matchesArr);
                } catch (BasketException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
            case UpdateMatches:{
                System.out.println("Match list updated Request ...");
                Ticket ticket=ProtoUtils.getTicket(request);
                try {
                    server.sendUpdatedList(ticket);
                    return ProtoUtils.createOkResponse();
                } catch (BasketException e) {
                    return ProtoUtils.createErrorResponse(e.getMessage());
                }
            }
        }
        return response;
    }

    private void sendResponse(BasketProtobufs.BasketResponse response) throws IOException {
        System.out.println("sending response "+response);
        response.writeDelimitedTo(output);
        //output.writeObject(response);
        output.flush();
    }
}
